// rcs-core/dest/index.js
"use strict";

var __createBinding = this && this.__createBinding || (Object.create ? function (o, m, k, k2) {
  if (k2 === undefined) k2 = k;
  Object.defineProperty(o, k2, {
    enumerable: true,
    get: function () {
      return m[k];
    }
  });
} : function (o, m, k, k2) {
  if (k2 === undefined) k2 = k;
  o[k2] = m[k];
});

var __setModuleDefault = this && this.__setModuleDefault || (Object.create ? function (o, v) {
  Object.defineProperty(o, "default", {
    enumerable: true,
    value: v
  });
} : function (o, v) {
  o["default"] = v;
});

var __importStar = this && this.__importStar || function (mod) {
  if (mod && mod.__esModule) return mod;
  var result = {};
  if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);

  __setModuleDefault(result, mod);

  return result;
};

var __importDefault = this && this.__importDefault || function (mod) {
  return mod && mod.__esModule ? mod : {
    "default": mod
  };
};

Object.defineProperty(exports, "__esModule", {
  value: true
});

const stats_1 = __importDefault(require("./stats"));

const replace_1 = __importDefault(require("./replace"));

const baseLibrary_1 = __importDefault(require("./baseLibrary"));

const fillLibraries_1 = __importDefault(require("./fillLibraries"));

const nameGenerator = __importStar(require("./nameGenerator"));

const selectorsLibrary_1 = __importDefault(require("./selectorsLibrary"));

const keyframesLibrary_1 = __importDefault(require("./keyframesLibrary"));

const cssVariablesLibrary_1 = __importDefault(require("./cssVariablesLibrary"));

const optimize_1 = __importDefault(require("./optimize"));

const generate_1 = __importDefault(require("./mapping/generate"));

const load_1 = __importDefault(require("./mapping/load"));

const generate_2 = __importDefault(require("./statistics/generate"));

const load_2 = __importDefault(require("./statistics/load"));

const extractFromHtml_1 = __importDefault(require("./helpers/extractFromHtml"));

const htmlToAst_1 = __importDefault(require("./helpers/htmlToAst"));

const allWarnings_1 = __importDefault(require("./allWarnings"));

exports.default = {
  stats: stats_1.default,
  replace: replace_1.default,
  baseLibrary: baseLibrary_1.default,
  fillLibraries: fillLibraries_1.default,
  nameGenerator,
  selectorsLibrary: selectorsLibrary_1.default,
  keyframesLibrary: keyframesLibrary_1.default,
  cssVariablesLibrary: cssVariablesLibrary_1.default,
  helpers: {
    htmlToAst: htmlToAst_1.default,
    extractFromHtml: extractFromHtml_1.default
  },
  mapping: {
    generate: generate_1.default,
    load: load_1.default
  },
  optimize: optimize_1.default,
  statistics: {
    generate: generate_2.default,
    load: load_2.default
  },
  warnings: allWarnings_1.default
};
module.exports = exports.default;
