// rcs-core/dest/nameGenerator.js
"use strict";

var __importDefault = this && this.__importDefault || function (mod) {
  return mod && mod.__esModule ? mod : {
    "default": mod
  };
};

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.useCustomGenerator = exports.NameGenerator = void 0;

const decimal_to_any_1 = __importDefault(require("decimal-to-any"));

const defaultGenerator = obj => (0, decimal_to_any_1.default)(obj.nameCounter, obj.alphabet.length, obj);

let customGenerator = null;

class NameGenerator {
  constructor(type = '') {
    this.nameCounter = 1;
    this.decToAnyOptions = {
      alphabet: 'etnrisouaflchpdvmgybwxk_jqz'
    };
    this.type = type;
  }

  setAlphabet(alphabet) {
    this.nameCounter = 1;
    this.decToAnyOptions.alphabet = alphabet;
  }

  generate(selector) {
    const genFunc = typeof customGenerator === 'function' ? customGenerator : defaultGenerator;
    const generatedName = genFunc({
      selector,
      type: this.type,
      alphabet: this.decToAnyOptions.alphabet,
      nameCounter: this.nameCounter
    });
    this.nameCounter += 1;
    return generatedName;
  }

  reset() {
    this.nameCounter = 1;
  }

}

exports.NameGenerator = NameGenerator;

const useCustomGenerator = generator => {
  customGenerator = generator;
};

exports.useCustomGenerator = useCustomGenerator;
