// rcs-core/dest/replace/regex.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
/* eslint-disable max-len */

exports.default = {
  //                          - matches all selectors beginning with . or # - e.g. .matching#alsomatch .matchmetoo
  //                         |            - matches backslash and something after
  //                         |           |           - matches everything after AGAIN, basically a copy from the beginning
  //                 (       v       )(  v  )(      v      )
  selectors: /(#|\.)[^\s.{)>+,~")[\\]+(\\.)?[^\s.{)>+,~")[]+/g,
  strings: /"\s*[\S\s]*?\s*"|'\s*[\S\s]*?\s*'/g,
  classAttributeSelectors: /\[\s*(class)\s*([*^$|~]?)=\s*("[^\]]*?"|'[^\]]*?'|[^\]]*)[^\]]*?\]/g,
  idAttributeSelectors: /\[\s*(id|for)\s*([*^$|~]?)=\s*("[^\]]*?"|'[^\]]*?'|[^\]]*)[^\]]*?\]/g,
  commonAttributeSelectors: /\[\s*(class|id|for)\s*([*^$|~]?)=\s*("[^\]]*?"|'[^\]]*?'|[^\]]*)[^\]]*?\]/g,
  keyframes: /@(-[a-z]*-)*keyframes\s+([a-zA-Z0-9_-]+)/g,
  cssVariables: /var\(\s*(--[^,\s]+?)(?:\s*,\s*(.+))?\s*\)/g,
  templateSelectors: /(class|for|id)+\s*=\s*(['"])((?:.(?!\2))*.?)\2/g,
  likelySelector: /[ #.=,()"'[\]]+/ // if it contains any of these chars, then the string is likely a selector

};
module.exports = exports.default;
